pkg/storage/stores/shipper/indexgateway/gateway.go

```diff
@@ -2,40 +2,213 @@ package indexgateway
+	// In case this function will return error we want to unregister the instance
+	// from the ring. We do it ensuring dependencies are gracefully stopped if they
+	// were already started.
+	defer func() {
+		if err == nil || g.subservices == nil {
+			return
+		}
+
+		if stopErr := services.StopManagerAndAwaitStopped(context.Background(), g.subservices); stopErr != nil {
+			level.Error(util_log.Logger).Log("msg", "failed to gracefully stop index gateway dependencies", "err", stopErr)
+		}
+	}()
+
+	if err := services.StartManagerAndAwaitHealthy(ctx, g.subservices); err != nil {
+		return errors.Wrap(err, "unable to start index gateway subservices")
+	}
+
+	// The BasicLifecycler does not automatically move state to ACTIVE such that any additional work that
+	// someone wants to do can be done before becoming ACTIVE. For the index gateway we don't currently
+	// have any additional work so we can become ACTIVE right away.
+	// Wait until the ring client detected this instance in the JOINING
+	// state to make sure that when we'll run the initial sync we already
+	// know the tokens assigned to this instance.
+	level.Info(util_log.Logger).Log("msg", "waiting until index gateway is JOINING in the ring")
+	if err := ring.WaitInstanceState(ctx, g.ring, g.ringLifecycler.GetInstanceID(), ring.JOINING); err != nil {
+		return err
+	}
+	level.Info(util_log.Logger).Log("msg", "index gateway is JOINING in the ring")
+
+	if err = g.ringLifecycler.ChangeState(ctx, ring.ACTIVE); err != nil {
+		return errors.Wrapf(err, "switch instance to %s in the ring", ring.ACTIVE)
+	}
+
+	// Wait until the ring client detected this instance in the ACTIVE state to
+	// make sure that when we'll run the loop it won't be detected as a ring
+	// topology change.
+	level.Info(util_log.Logger).Log("msg", "waiting until index gateway is ACTIVE in the ring")
+	if err := ring.WaitInstanceState(ctx, g.ring, g.ringLifecycler.GetInstanceID(), ring.ACTIVE); err != nil {
+		return err
+	}
+	level.Info(util_log.Logger).Log("msg", "index gateway is ACTIVE in the ring")
+
+	return nil
+}
+
+// running implements the Lifecycler interface and is one of the lifecycle hooks.
+//
+// Only invoked if the Index Gateway is in ring mode.
+func (g *Gateway) running(ctx context.Context) error {
+	t := time.NewTicker(ringCheckPeriod)
+	defer t.Stop()
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case err := <-g.subservicesWatcher.Chan():
+			return errors.Wrap(err, "running index gateway subservice failed")
+		case <-t.C:
+			continue
+			// TODO: should we implement CAS check?
+		}
+	}
 }
 
-func NewIndexGateway(indexQuerier IndexQuerier) *gateway {
-	g := &gateway{
+// stopping implements the Lifecycler interface and is one of the lifecycle hooks.
+//
+// Only invoked if the Index Gateway is in ring mode.
+func (g *Gateway) stopping(_ error) error {
+	level.Debug(util_log.Logger).Log("msg", "stopping index gateway")
+	g.shipper.Stop()
```

Comment:
I think `g.shipper` should be stopped last just for safety since the instance would still be active in the ring which would make the index gateways to keep getting queries.

Reply:
I added a defer to `g.shipper.Stop()`, do you think it is enough?